DDS-738 serialize project creator parent

dmlond opened the pull request.

dmlond committed:
UserSearchDocumentSerialalizerSpec expects UserSearchDocumentSerializer to serialize a User
object with its id, username, email, first_name, last_name

UserSearchDocumentSerializer meets spec
DDS-746 #resolve

dmlond committed:
ProjectSearchDocumentSerializerSpec expects ProjectSearchDocumentSerializer to serialize a Project
with just its id
ProjectSearchDocumentSerializer meets spec

DDS-745 #resolve

dmlond committed:
casey wanted the name too

dmlond committed:
DataFileSearchDocumentSerializerSpec expects DataFileSearchDocumentSerializer to have one project
serialized with the ProjectSearchDocumentSerializer

DataFileSearchDocumentSerializer meets spec

dmlond committed:
Warning! Major Refactor.
Moved all of the SearchDocumentSerializers into a seperate Search:: namespace. This allows us to follow general conventions for naming
serializers for both API serialization (not in the Search:: namespace), and Elasticsearch serialization (in the Search:: namespace).

If an object should be serialized for direct usage, e.g. returned by the API, or seralized to a search index in Elasticsearch, then its name
should reflect that with the object name and Serializer. FolderSerializer is meant to serialize a Folder to return in the API, while Search::FolderSerializer
is meant to be serialized to the folders indices in Elasticsearch.

If an object should be serialized inside of another object, especially when the object is meant to be 'summarized' by removing attributes, associations, etc.
that would normally be returned for the full object, it should be serialized with the object name and SummarySerializer. FolderSummarySerializer is meant to
add a summary of a Folder to the serialization of another object returned by the API, while Search::FolderSummarySerializer is meant to add a summary of
a Folder to the Elasticsearch index for another objects document.

dmlond committed:
renamed:
- Search::ProjectSerializer -> Search::ProjectSummarySerializer
- Search::TagSerializer -> Search::TagSummarySerializer
- Search::UserSerializer -> Search::UserSummarySerializer

as these are meant to be used inside Search::DataFileSerializer and Search::FolderSerializer

dmlond committed:
renamed incorrectly in the previous refactor.
ALSO bugfix. It was testing a data_file factory instead of a folder factory!

dmlond committed:
Search::FolderSummarySerializerSpec expects Search::FolderSummarySerializer to serialize a Folder
with its id and name
Search::FolderSummarySerializer meets spec

dmlond committed:
Search::DataFileSerializerSpec and Search::FolderSerializerSpec expect Search::DataFileSerializer and
Search::FolderSerializer to serialize their parent with a Search::FolderSummarySerializer

dmlond committed:
Search::DataFileSerializer and Search::FolderSerializer are expected to
serialize their creator using the Search::UserSummarySerializer

DataFile creator returns the user attached to the creation audit for the current_file_version
Folder creator returns the user attached to the creation audit for the Folder

dmlond committed:
Search::*Serializer specs do not test that the creator returned is the expected creator

DataFileSpec expects DataFile creator to return the user attached to the creation audit for the current_file_version
FolderSpec expects Folder creator to return the user attached to the creation audit for the Folder

dmlond committed:
creator test was in the elasticsearch section, pulled out

dmlond committed:
refactored the elasticsearch spec tests into two shared_examples in search_shared_examples

dmlond committed:
added parent and creator mappings to data_file and folder
added project mappings to data_file

dmlond committed:
added a FEATURE_TOGGLED deployment command to drop and recreate all the elasticsearch indices if RECREATE_SEARCH_MAPPINGS
environment variable exists and is nonzero length

dmlond committed:
bugfix

dmann commented on app/models/folder.rb:
```diff
@@ -33,7 +33,7 @@ def is_deleted=(val)
 
   def as_indexed_json(options={})
-    FolderSearchDocumentSerializer.new(self).as_json
+    Search::FolderSerializer.new(self).as_json
```
Namespacing FTW! 👍

dmann commented on app/models/data_file.rb:
```diff
@@ -83,6 +83,27 @@ def as_indexed_json(options={})
   end
+
+  def creator
+    current_file_version.audits.find_by(action: "create").user
```
Might want to put a guard clause around this in case current_file_version, .audits, or .find_by(action: "create") return nil.

dmlond replied:
👍

dmann commented on app/models/folder.rb:
```diff
@@ -43,6 +43,24 @@ def as_indexed_json(options={})
   end
+
+  def creator
+    audits.find_by(action: "create").user
```
A guard clause would be good incase audits.find_by returns nil. Though, maybe it'll throw a Record Not Found exception, which could be better/worse?  :\

dmlond replied:
👍

dmann commented on app/serializers/search/data_file_serializer.rb:
```diff
@@ -0,0 +1,27 @@
+    object.is_deleted?
+  end
+
+  def meta
```
Could the `meta` method possibly work as a has_many? Not a criticism, I'm honestly curious.

dmlond replied:
I dont know? It certainly doesnt lend itself to it. It wants a meta hash, with a series of key (meta_template.template.name): Hash with series of key (meta_property.property.key): meta_property.value entries. These do not align nicely with the underlying object structures

dmann commented on circle.yml:
```diff
@@ -46,6 +46,9 @@ deployment:
     - heroku run --exit-code rake auth_service:set_default --app 'dukeds-dev'
+    - "[[ -z ${RECREATE_SEARCH_MAPPINGS} ]] || heroku run --exit-code rake elasticsearch:index:drop --app 'dukeds-dev'"
+    - "[[ -z ${RECREATE_SEARCH_MAPPINGS} ]] || heroku run --exit-code rake elasticsearch:index:create  --app 'dukeds-dev'"
+    - "[[ -z ${RECREATE_SEARCH_MAPPINGS} ]] || heroku run --exit-code rake elasticsearch:index:index_documents  --app 'dukeds-dev'"
```
Future us should create a single rake task to wrap the 3 elasticsearch tasks needed for recreating search mappings.

dmlond replied:
good idea

dmann commented on spec/models/data_file_spec.rb:
```diff
@@ -245,48 +245,75 @@
+    it {
+      expect(subject.current_file_version).not_to be_nil
+      expect(subject.current_file_version.audits).not_to be_empty
+      expect(subject.current_file_version.audits.find_by(action: 'create')).not_to be_nil
```
These 3 expectations (lines 257-9) could be broken out into their own contexts, which will ultimately test the guard clause in your creator method.

dmlond replied:
👍

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -141,39 +141,55 @@
-    # TODO, when we move to asynchronous indexing, remove this and replace with
-    # a test to ensure that jobs are created on create, update, delete
-    it { expect(described_class).to include(Elasticsearch::Model::Callbacks) }
+  describe 'creator' do
```
Change to `describe '#creator' do`, we should try to follow the naming convention at http://betterspecs.org/#describe
So, `#` for instance methods and `::` for class methods. (I'm moving away from using the `.` as I've found that it's not explicit enough)

dmlond replied:
thanks. will do this

dmlond replied:
👍

dmann commented on spec/models/folder_spec.rb:
```diff
@@ -141,39 +141,55 @@
-    end
+    it { is_expected.to respond_to :creator }
+    it {
+      expect(subject.audits.find_by(action: 'create')).not_to be_nil
```
See the above comment regarding contexts and guard clauses.

dmlond replied:
👍

dmann reviewed (changes requested):
Some changes and a few questions. Good work over all!

dmlond committed:
added guard clause to #creator
tested #creator with expectations for nil current_version, nil current_version creation audit, and existing current_version creation audit
resolves code review comment

dmlond committed:
added guard clause for #creator
tests for #creator test guard clause with contexts for nil and existing creation_audit
resolves code review comment
